enterprise/cypher/compiled-expressions/src/main/java/org/neo4j/cypher/internal/runtime/compiled/expressions/CypherMath.java

```diff
@@ -0,0 +1,238 @@
+import org.neo4j.values.storable.TimeValue;
+import org.neo4j.values.storable.Value;
+import org.neo4j.values.storable.Values;
+import org.neo4j.values.virtual.ListValue;
+import org.neo4j.values.virtual.MapValue;
+import org.neo4j.values.virtual.PathValue;
+import org.neo4j.values.virtual.VirtualNodeValue;
+import org.neo4j.values.virtual.VirtualRelationshipValue;
+import org.neo4j.values.virtual.VirtualValues;
+
+import static org.neo4j.values.storable.Values.NO_VALUE;
+import static org.neo4j.values.storable.Values.doubleValue;
+import static org.neo4j.values.storable.Values.stringValue;
+
+/**
+ * This class contains static helper math methods used by the compiled expressions
+ */
+@SuppressWarnings( "unused" )
+public final class CypherMath
+{
+    private CypherMath()
+    {
+        throw new UnsupportedOperationException( "Do not instantiate" );
+    }
+
+    //TODO this is horrible spaghetti code, we should push most of this down to AnyValue
+    public static AnyValue add( AnyValue lhs, AnyValue rhs )
+    {
+        if ( lhs == NO_VALUE || rhs == NO_VALUE )
+        {
+            return NO_VALUE;
+        }
+        if ( lhs instanceof NumberValue && rhs instanceof NumberValue )
+        {
+            return ((NumberValue) lhs).plus( (NumberValue) rhs );
+        }
+        //List addition
+        //arrays are same as lists when it comes to addition
+        if ( lhs instanceof ArrayValue )
+        {
+            lhs = VirtualValues.fromArray( (ArrayValue) lhs );
+        }
+        if ( rhs instanceof ArrayValue )
+        {
+            rhs = VirtualValues.fromArray( (ArrayValue) rhs );
+        }
+
+        boolean lhsIsListValue = lhs instanceof ListValue;
+        if ( lhsIsListValue && rhs instanceof ListValue )
+        {
+            return VirtualValues.concat( (ListValue) lhs, (ListValue) rhs );
+        }
+        else if ( lhsIsListValue )
+        {
+            return ((ListValue) lhs).append( rhs );
+        }
+        else if ( rhs instanceof ListValue )
+        {
+            return ((ListValue) rhs).prepend( lhs );
+        }
+
+        // String addition
+        if ( lhs instanceof TextValue && rhs instanceof TextValue )
+        {
+            return stringValue( ((TextValue) lhs).stringValue() + ((TextValue) rhs).stringValue() );
+
+        }
+        else if ( lhs instanceof TextValue )
+        {
+            if ( rhs instanceof Value )
+            {
+                // Unfortunately string concatenation is not defined for temporal and spatial types, so we need to
+                // exclude them
+                if ( !(rhs instanceof TemporalValue || rhs instanceof DurationValue || rhs instanceof PointValue) )
+                {
+                    return stringValue( ((TextValue) lhs).stringValue() + ((Value) rhs).prettyPrint() );
+                }
+                else
+                {
+                    //TODO this seems wrong but it is what we currently do in compiled runtime
```

Comment:
Is this TODO cardified so we don't forget?

Reply:
I misread the code in `CompiledMathHelper`, it doesn't seem as wrong to me anymore. Removed the TODO